commands/debug/debug.go

```diff
@@ -0,0 +1,222 @@
+// This file is part of arduino-cli.
+//
+// Copyright 2020 ARDUINO SA (http://www.arduino.cc/)
+//
+// This software is released under the GNU General Public License version 3,
+// which covers the main part of arduino-cli.
+// The terms of this license can be found at:
+// https://www.gnu.org/licenses/gpl-3.0.en.html
+//
+// You can be released from the requirements of the above licenses by purchasing
+// a commercial license. Buying such a license is mandatory if you want to
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package debug
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/arduino/arduino-cli/arduino/cores"
+	"github.com/arduino/arduino-cli/arduino/cores/packagemanager"
+	"github.com/arduino/arduino-cli/arduino/sketches"
+	"github.com/arduino/arduino-cli/commands"
+	"github.com/arduino/arduino-cli/executils"
+	dbg "github.com/arduino/arduino-cli/rpc/debug"
+	"github.com/arduino/go-paths-helper"
+	"github.com/arduino/go-properties-orderedmap"
+	"github.com/sirupsen/logrus"
+)
+
+// Debug command launches a debug tool for a sketch.
+// It also implements streams routing:
+// gRPC In -> tool stdIn
+// grpc Out <- tool stdOut
+// grpc Out <- tool stdErr
+// It also implements tool process lifecycle management
+func Debug(ctx context.Context, req *dbg.DebugConfigReq, inStream io.Reader, out io.Writer) (*dbg.DebugResp, error) {
+
+	// Get tool commandLine from core recipe
+	pm := commands.GetPackageManager(req.GetInstance().GetId())
+	commandLine, err := getCommandLine(req, pm)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get command line for tool: %s", err)
+	}
+
+	// Run Tool
+	cmd, err := executils.Command(commandLine)
+	if err != nil {
+		return nil, fmt.Errorf("cannot execute debug tool: %s", err)
+	}
+
+	// Get stdIn pipe from tool
+	in, err := cmd.StdinPipe()
+	if err != nil {
+		fmt.Printf("%v\n", err)
+		return &dbg.DebugResp{Error: err.Error()}, nil
+	}
+	defer in.Close()
+
+	// Merge tool StdOut and StdErr to stream them in the io.Writer passed stream
+	cmd.Stdout = out
+	cmd.Stderr = out
+
+	// Start the debug command
+	if err := cmd.Start(); err != nil {
+		fmt.Printf("%v\n", err)
+		return &dbg.DebugResp{Error: err.Error()}, nil
+	}
+
+	go func() {
+		// Copy data from passed inStream into command stdIn
+		io.Copy(in, inStream)
+		// In any case, try process termination after a second to avoid leaving
+		// zombie process.
+		time.Sleep(time.Second)
+		cmd.Process.Kill()
```

Comment:
Yes, it happens when the pipe is not gracefully terminated (for example if we kill the client with CTRL-C), in this case we do not receive the EOF and the process is left zombie.
We may catch this error and kill the process only in that case, but in the end we opted to kill the process anyway (after a second of delay) to avoid unforeseen cases like this one.
If the process terminates normally, the `Kill` will just fail silently.

Reply:
Basically the only case of graceful `gdb` process close is when a `quit` command is sent. 
Other cases has to be managed by the cli.

Reply:
I'm not sure I understand this entirely. If the process is terminated ungracefully, why kill it (it will be already dead). Also, zombie processes are processes which are actually dead, but are still waiting for their parent to reap them (e.g. read their exit status). This should be done using a "wait" call, which I suppose kill might do internally, but which is also done a few lines below, so I wonder why this kill is really needed, then.

Reply:
The process is not zombie, is just "hanging" (this is a mistake in the comment).
In other words if `gdb` doesn't receive the `quit` command it just sits there: it's not dead, it's not zombie... it's just running waiting for input.

If something goes wrong on the IDE side or on the OpenOCD side, for any reason, gdb will not receive `quit` and will not close itself gracefully, the only way out is to kill it. But we have no way to understand if everything went smooth unless we inspect the stdin/out traffic...

That's the reason why we kill it anyway. We could check if the process is still running and kill it only if still running, but it will just complicate things without a real benefit.

> This should be done using a "wait" call, which I suppose kill might do internally

No kill doesn't wait, the `wait` syscall is made here https://github.com/arduino/arduino-cli/pull/590/files#diff-ec656dd0bf60c33c4e4144e7d5e9508aR86 outside of the go-routine, basically it waits for the subprocess to terminate to finish the external debug call.